### ISSUE TITLE
EventEmitter memory leak when sockets are heavily reused

### ISSUE THREAD

**Issue by martonw** [tomas/needle#161]
Hi - 

the issue looks like when I am using HTTP keep-alive (which is the default anyway) and there if a heavy outgoing HTTP traffic, the (HTTP) socket objects are being reassigned to new HTTP requests. However, at some point there is an event listener which is attached to the socket object for every request, and eventually node starts displaying the following error messages:

(node) warning: possible EventEmitter memory leak detected. 11 end listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at Socket.addListener (events.js:179:15)
    at Socket.Readable.on (_stream_readable.js:671:33)
    at Socket.once (events.js:204:8)
    at ClientRequest.<anonymous> (/home/marci/impress/ctl-perftest/node_modules/needle/lib/needle.js:591:12)
    at ClientRequest.g (events.js:199:16)
    at ClientRequest.emit (events.js:129:20)
    at tickOnSocket (_http_client.js:493:7)
    at _http_client.js:504:7
    at process._tickCallback (node.js:355:11)

**Comment by martonw** [comment]
I submitted a merge request in the meantime to fix this. A bit of a background on the issue:

when needle creates a http(s) request, it also creates a listener on the underlying socket object to detect if the socket has been closed:
```javascript
  // handle socket 'end' event to ensure we don't get delayed EPIPE errors.
  request.once('socket', function(socket) {
    socket.once('end', on_socket_end.bind(socket));
  })
```
and it suppose to remove the listener once the request is complete like this:
```javascript
request.socket.removeListener('end', on_socket_end);
```

Now the problem is that the removeListener will not work as expected, and as a result, the socket 'end' listerer will never be removed. However, if you are using keep-alive, which is the default mode for http sockets, node will reuse the open sockets if the next request will come sooner than the keepAlive timeout - which is quite typical scenario for high load sites. However, in this case, the `socket.once('end', ...)` will be called each time a socket is reused for a new request, and eventually it will register 10+ listeners on this event, resulting in the above node warning message (and - indeed, a memory leak). 

Why the `removeListener` is not working as expected? 
1. The `EventEmitter.once` will actually wrap the passed function into another function, so calling `EventEmitter.removeListener` with the same function will not remove it. So `EventEmitter.once` will not work the same way as `EventEmitter.on` does. In this case, however, I see no problem using on() rather than once() as the 'end' event will fire only once anyway.
2. Similarly as above, `myfunction` != `myfunction.bind(someObject)`, as the latter will again create a sort of wrapping around your original function, so you need to find another way to make sure the exact same function is passed to `EventEmitter.on` and `EventEmitter.removeListener`.

**Comment by martonw** [comment]
Hi @tomas - did you have a chance to review the latest post + proposed CR? 

Thanks.

**Comment by tomas** [comment]
Not really. I hope I can take a look later today.

**Comment by tomas** [comment]
Closed thanks to your PR #162.

Thanks again and sorry for the long delay. :)
